### PR TITLE
Charts improvements

### PR DESCRIPTION
**Pull request by rsanchezsaez**
Replaces [PR #385](https://github.com/ResearchKit/ResearchKit/pull/385).

---

Several improvements to the charts:

*Graph Charts*
- Add optional delegate method to provide colors for secondary graphs on multi-plot graphs ([Issue #379](https://github.com/ResearchKit/ResearchKit/issues/379)).
- Add optional delegate method to allow to select which graph to scrub over on multi-plot graphs  ([Issue #379](https://github.com/ResearchKit/ResearchKit/issues/379)).
- Fix scrubbing animation glitch when you scrub over one graph edge, release finger, and start scrubbing over the opposite edge.

*Pie Chart*
- Use shades of current `tintColor` (instead of shades of black) when no colors are supplied to the pie chart view.
- Fix animation not working if `showsPercentageLabels` was set to `NO`.

*All*
- Properties with default values can be reset to their default value if set to `nil` (and declare them as `null_resetable`).
- Documentation improvements.
- Other minor fixes.

### CODE REVIEW

**Comment by rsanchezsaez** [comment]
Done. Let me know if you have any feedback. I've done several tests and nothing seems to be broken (I've not tested accessibility, though).

File: ResearchKit/Charts/ORKGraphChartView.m [hunk 1]
```diff
@@ -693,14 +748,20 @@ - (void)updateScrubberLineAccessories:(CGFloat)xPosition {
...
+        [self setScrubberAccessoryViewsHidden:NO];
+    }
 }
 
-- (CGFloat)snappedXPosition:(CGFloat)xPosition {
+- (CGFloat)snappedXPosition:(CGFloat)xPosition plotIndex:(NSInteger)plotIndex {
     CGFloat numberOfXAxisPoints = self.numberOfXAxisPoints;
     CGFloat widthBetweenPoints = CGRectGetWidth(_plotView.frame) / numberOfXAxisPoints;
-    for (NSUInteger positionIndex = 0; positionIndex < ((NSArray *)_dataPoints[0]).count; positionIndex++) {
+    for (NSUInteger positionIndex = 0; positionIndex < ((NSArray *)_dataPoints[plotIndex]).count; positionIndex++) {
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKGraphChartView.m]: Declare `NSInteger positionCount = ((NSArray *)_dataPoints[plotIndex]).count;` and use it in the for loop?

File: ResearchKit/Charts/ORKGraphChartView.m [hunk 2]
```diff
@@ -713,53 +774,50 @@ - (CGFloat)snappedXPosition:(CGFloat)xPosition {
...
+        NSInteger positionIndex = 0;
+        NSInteger numberOfXAxisPoints = self.numberOfXAxisPoints;
         for (positionIndex = 0; positionIndex < (numberOfXAxisPoints - 1); positionIndex++) {
             CGFloat xAxisPointValue = xAxisPoint(positionIndex, numberOfXAxisPoints, _plotView.bounds.size.width);
             if (xAxisPointValue == xPosition) {
                 break;
             }
         }
-        value = ((ORKRangedPoint *)_dataPoints[0][positionIndex]).maximumValue;
+        value = ((ORKRangedPoint *)_dataPoints[plotIndex][positionIndex]).maximumValue;
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKGraphChartView.m]: Would it be better to create a helper method to retrieve `ORKRangedPoint` from the nested array? So we don't need to cast it each time.

File: ResearchKit/Charts/ORKGraphChartView.m (hunk 1 above)
Comment (rsanchezsaez) [inline ResearchKit/Charts/ORKGraphChartView.m]: Sure thing. I actually think the compiler optimizes this anyway, but it's easier to read over two lines.

File: ResearchKit/Charts/ORKLineGraphChartView.m [hunk 3]
```diff
@@ -159,29 +158,29 @@ - (void)layoutLineLayersForPlotIndex:(NSInteger)plotIndex {
...
+        NSInteger previousValidIndex = [self previousValidPointIndexForPointIndex:pointIndex plotIndex:plotIndex];
+        NSInteger nextValidIndex = [self nextValidPointIndexForPointIndex:pointIndex plotIndex:plotIndex];
         
         CGFloat x1 = xAxisPoint(previousValidIndex, numberOfXAxisPoints, viewWidth);
         CGFloat x2 = xAxisPoint(nextValidIndex, numberOfXAxisPoints, viewWidth);
         
-        CGFloat y1 = ((ORKRangedPoint *)self.dataPoints[0][previousValidIndex]).minimumValue;
-        CGFloat y2 = ((ORKRangedPoint *)self.dataPoints[0][nextValidIndex]).minimumValue;
+        CGFloat y1 = ((ORKRangedPoint *)self.dataPoints[plotIndex][previousValidIndex]).minimumValue;
+        CGFloat y2 = ((ORKRangedPoint *)self.dataPoints[plotIndex][nextValidIndex]).minimumValue;
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKLineGraphChartView.m]: Convert this two to use helper method?

File: ResearchKit/Charts/ORKLineGraphChartView.m [hunk 4]
```diff
@@ -217,11 +216,23 @@ - (CGFloat)canvasYPointForXPosition:(CGFloat)xPosition {
     return canvasYPosition;
 }
 
+- (NSInteger)nextValidPointIndexForPointIndex:(NSInteger)pointIndex plotIndex:(NSInteger)plotIndex {
+    NSUInteger validPosition = pointIndex;
+    
+    while (validPosition < (((NSArray *)self.dataPoints[plotIndex]).count - 1)) {
+        if (((ORKRangedPoint *)self.dataPoints[plotIndex][validPosition]).maximumValue != ORKCGFloatInvalidValue) {
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKLineGraphChartView.m]: And this.

File: ResearchKit/Charts/ORKLineGraphChartView.m [hunk 5]
```diff
@@ -217,11 +216,23 @@ - (CGFloat)canvasYPointForXPosition:(CGFloat)xPosition {
...
+    return validPosition;
+}
 
-- (NSInteger)previousValidPositionIndexForPosition:(NSInteger)positionIndex {
-    NSInteger validPosition = positionIndex - 1;
+- (NSInteger)previousValidPointIndexForPointIndex:(NSInteger)pointIndex plotIndex:(NSInteger)plotIndex {
+    NSInteger validPosition = pointIndex - 1;
     while (validPosition > 0) {
-        if (((ORKRangedPoint *)self.dataPoints[0][validPosition]).minimumValue != ORKCGFloatInvalidValue) {
+        if (((ORKRangedPoint *)self.dataPoints[plotIndex][validPosition]).minimumValue != ORKCGFloatInvalidValue) {
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKLineGraphChartView.m]: This

File: ResearchKit/Charts/ORKDiscreteGraphChartView.m [hunk 6]
```diff
@@ -65,7 +65,7 @@ - (void)updateLineLayersForPlotIndex:(NSInteger)plotIndex {
         ORKRangedPoint *dataPointValue = self.dataPoints[plotIndex][i];
```
Comment (YuanZhu-apple) [inline ResearchKit/Charts/ORKDiscreteGraphChartView.m]: Use helper method instead?

**Comment by rsanchezsaez** [comment]
Replaced the remaining `self.dataPoint` subclass access by helper method.

**Comment by YuanZhu-apple** [comment]
r=me.

**Pull request merged**